### PR TITLE
Implementation of an endpoint for fetching requisitions

### PR DESCRIPTION
**Pull request by honourableola**
Endpoint for getting requisitions implemented. The endpoint takes a flag (usePaging) which is used to determine the Query handler to invoke. Two handlers were implemented, one for paginated results and the other for non=paginated results.

### CODE REVIEW

File: src/Application/Queries/GetPaginatedRequisitions.cs [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+{
+    public class GetPaginatedRequisitions
+    {
+        public record Query : PageRequest, IRequest<PaginatedList<RequisitionResponse>>
+        {
+            public DateTime? StartDate { get; set; }
+            public DateTime? EndDate { get; set; }
+            public string? ExpenseHead { get; set; }
+            public string? Department { get; set; }
+            public string? StatusFilter { get; set; } //Assumption is that FE sends comma separated values
```
Comment (adeyemi-malik) [inline src/Application/Queries/GetPaginatedRequisitions.cs]: please remove assuptions in your implemetation.. you have to be solid on what the contract is and have frontend adhere to the contract

File: src/Application/Queries/GetPaginatedRequisitions.cs [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+            public HashSet<int> RequisitionStatuses => string.IsNullOrWhiteSpace(StatusFilter) ? new HashSet<int>() : StatusFilter.Split(',').Select(int.Parse).ToHashSet();
```
Comment (adeyemi-malik) [inline src/Application/Queries/GetPaginatedRequisitions.cs]: i think the statuses are better defined as enums

File: src/Application/Queries/GetPaginatedRequisitions.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+            public string? TypeFilter { get; set; } //Assumption is that FE sends comma separated values
```
Comment (adeyemi-malik) [inline src/Application/Queries/GetPaginatedRequisitions.cs]: again, please remove assumptions in implementation

File: src/Application/Queries/GetPaginatedRequisitions.cs [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+            public HashSet<int> RequisitionTypes => string.IsNullOrWhiteSpace(TypeFilter) ? new HashSet<int>() : TypeFilter.Split(',').Select(int.Parse).ToHashSet();
```
Comment (adeyemi-malik) [inline src/Application/Queries/GetPaginatedRequisitions.cs]: same here.. better to have it defined as enums and not int

File: src/Application/Queries/GetPaginatedRequisitions.cs [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+            {
+                var pageRequest = new PageRequest
+                {
+                    Page = request.Page,
+                    PageSize = request.PageSize,
+                    IsAscending = request.IsAscending,
+                    Keyword = request.Keyword,
+                    SortBy = request.SortBy
+                };
+                var requisitions = await _requisitionRepository.GetAllAsync(pageRequest, request.StartDate, request.EndDate, request.ExpenseHead, request.Department, request.RequisitionStatuses, request.RequisitionTypes);
```
Comment (adeyemi-malik) [inline src/Application/Queries/GetPaginatedRequisitions.cs]: why do we have a `GetAllAsync` method taking pagination parameter if it's truly getting all

File: src/Infrastructure/Persistence/Repositories/RequisitionRepository.cs [hunk 6]
```diff
@@ -27,7 +30,67 @@ public async Task<Requisition> UpdateAsync(Requisition requisition)
...
+                }
+            }
+
+            // Additional filters based on method parameters
+            if (startDate.HasValue && startDate.Value != DateTime.MinValue)
+                query = query.Where(r => r.RequestedDate.Date >= startDate.Value.Date);
+            if (endDate.HasValue && endDate.Value != DateTime.MinValue)
+                query = query.Where(r => r.RequestedDate.Date <= endDate.Value.Date);
+            if (!string.IsNullOrWhiteSpace(expenseHead))
+                query = query.Where(r => EF.Functions.Like(r.ExpenseHead.ToLower(), expenseHead.ToLower()));
```
Comment (adeyemi-malik) [inline src/Infrastructure/Persistence/Repositories/RequisitionRepository.cs]: why do we have `.ToLower` , db queries are mostly case-insensitive.

File: src/Application/Queries/GetPaginatedRequisitions.cs (hunk 1 above)
Comment (honourableola) [reply src/Application/Queries/GetPaginatedRequisitions.cs]: Removed

File: src/Application/Queries/GetPaginatedRequisitions.cs (hunk 2 above)
Comment (honourableola) [reply src/Application/Queries/GetPaginatedRequisitions.cs]: Comment addressed, thanks

File: src/Application/Queries/GetPaginatedRequisitions.cs (hunk 3 above)
Comment (honourableola) [reply src/Application/Queries/GetPaginatedRequisitions.cs]: Removed.

File: src/Application/Queries/GetPaginatedRequisitions.cs (hunk 4 above)
Comment (honourableola) [reply src/Application/Queries/GetPaginatedRequisitions.cs]: Addressed

File: src/Application/Queries/GetPaginatedRequisitions.cs (hunk 5 above)
Comment (honourableola) [reply src/Application/Queries/GetPaginatedRequisitions.cs]: Resolved

File: src/Infrastructure/Persistence/Repositories/RequisitionRepository.cs (hunk 6 above)
Comment (honourableola) [reply src/Infrastructure/Persistence/Repositories/RequisitionRepository.cs]: Comment addressed

**Review by NurudeenAdeyemi — DISMISSED** [review]
I think you can use a single query handler for the two. What do you think

**Comment by honourableola** [comment]
> I think you can use a single query handler for the two. What do you think

It's for simplicity sir, since getting all does not necessarily need to have Paginated response. Otherwise putting the two together in a handler forces Paginated response on GetAll

**Pull request merged**